src/Compilers/Core/Portable/SourceGeneration/GeneratorSyntaxWalker.cs

```diff
@@ -33,11 +32,14 @@ public void VisitWithModel(Lazy<SemanticModel>? model, SyntaxNode node)
             _semanticModel = null;
         }
 
-        public override void Visit(SyntaxNode node)
+        public void Visit(SyntaxNode node)
         {
-            Debug.Assert(_semanticModel is object && _semanticModel.Value.SyntaxTree == node.SyntaxTree);
-            _syntaxReceiver.OnVisitSyntaxNode(new GeneratorSyntaxContext(node, _semanticModel, _syntaxHelper));
-            base.Visit(node);
+            Debug.Assert(_semanticModel is object);
+            foreach (var child in node.DescendantNodesAndSelf())
+            {
+                Debug.Assert(_semanticModel.Value.SyntaxTree == child.SyntaxTree);
```

Comment:
>Debug.Assert(_semanticModel.Value.SyntaxTree == child.SyntaxTree); [](http://example.com/codeflow?start=16&length=66)

Can descendant nodes belong to a different tree? Otherwise, consider asserting only for ```node``` and before the loop.

Reply:
I don't have the context to answer that, but the assertion behavior should be equivalent with my changes.